### PR TITLE
server ring_duration failed to wait and send 200

### PR DESCRIPTION
**Pull request by hoilchoi**
when testing with UAC and UAS with uas configured with ring_duration, UAS 1) did not wait the duration and 2) did not send 200 OK resulting UAC to send call cancel.  This commit is to fix it so that UAS would wait for the ring_duration period and successfully send 200 OK back to UAC.

### CODE REVIEW

File: src/voip_patrol/voip_patrol.cc [hunk 1]
```diff
@@ -454,6 +454,12 @@ void TestCall::onCallState(OnCallStateParam &prm) {
 		test->call_id = getId();
 		test->sip_call_id = ci.callIdString;
 	}
+	if (test && (ci.lastStatusCode == 180 || ci.lastStatusCode == 183)) {
+		if (test->ring_duration > 0) {
+			LOG(logINFO) <<__FUNCTION__<<": Waiting "<<test->ring_duration <<" seconds ring_duration";
+			pj_thread_sleep(test->ring_duration * 1000);
```
Comment (igorolhovskiy) [inline src/voip_patrol/voip_patrol.cc:460]: Just to mention, that string is a blocking operation for all `voip_partol` execution and will block any other operations, that should be executed within `ring_duration` time.

File: src/voip_patrol/voip_patrol.cc [hunk 2]
```diff
@@ -637,20 +643,22 @@ void TestAccount::onIncomingCall(OnIncomingCallParam &iprm) {
...
 	call->answer(prm_100);
 
-	LOG(logINFO) <<__FUNCTION__<<"code:" << code <<" reason:"<< reason;
+	LOG(logINFO) <<__FUNCTION__<<": code:" << code <<" reason:"<< reason;
+	if (ring_duration > 0) {
+		prm.statusCode = PJSIP_SC_RINGING;
+		if (early_media) {
+			prm.statusCode = PJSIP_SC_PROGRESS;
+		}
+		call->answer(prm);
```
Comment (igorolhovskiy) [inline src/voip_patrol/voip_patrol.cc:652]: Isn't it a duplication for `call->answer()` that is executed below?

**Comment by igorolhovskiy** [comment]
@hoilchoi , thanks for the PR, but it still a bit unclear to me what is not working in your case. You mean in `accept` action `ring_duration` is not respected?

And for CANCEL - 200 OK sequence, you're saying, that 200 OK after CANCEL by UAC is not sent? Am I correct?

Can you pls share your XML what you're using?

Thanks!

File: src/voip_patrol/voip_patrol.cc (hunk 1 above)
Comment (hoilchoi) [reply src/voip_patrol/voip_patrol.cc:460]: I see, wondering if that is an issue tho during requested `ring_duration` time?  
does it block incoming CANCEL requests from UAC for example?

File: src/voip_patrol/voip_patrol.cc (hunk 2 above)
Comment (hoilchoi) [reply src/voip_patrol/voip_patrol.cc:652]: this one sets answer with either 180/183.  
one below sets answer with 200. 
without this, 200 was not sent and 180/183 became final response.

File: src/voip_patrol/voip_patrol.cc (hunk 1 above)
Comment (hoilchoi) [reply src/voip_patrol/voip_patrol.cc:460]: Yes it is the case, I'll rethink this.. thanks you!

**Comment by jchavanton** [comment]
Hi @hoilchoi hope you are well, adding `pj_thread_sleep` will sleep the main thread worker.
This will work just fine if you only are running one test or tests that are all synchronized somehow.


We need to add something in the action `do_wait`

in fact there is already some logic to handle the ring duration, now sure what is not working as expected, I will have to look more closely at it.


See this line in `do_action`
```
} else if (test->ring_duration > 0 && ci.totalDuration.msec >= (test->ring_duration * 1000 + test->response_delay)) {
```

**Comment by hoilchoi** [comment]
Hi @jchavanton , 

Hope you are doing well too.  
Ihor thankfully pointed it out and I realized that.  Unfortunuately, didn't have time to look further as I wished.  
I'm going to close this for now and start over when I have chance to. 

Thank you,

**Pull request closed without merging**